### PR TITLE
Addition of image requirements metadata

### LINKED ISSUE #9408: Add image requirements metadata
Adding additional image metadata that indicates the requirements the image has, so that LXD can inspect it and check the system and settings for compatibility.

This would be useful for (at least) the following two scenarios:

1. VM images that require secure boot be disabled.
2. Container images that use systemd and require the host to have a supported cgroup configuration.

If LXD could access the requirements of the image it could then detect if there are issues that would prevent the instance from working and flag it to the end user.

Otherwise in both scenarios above it is very difficult for the end user to diagnose the fault.

### PR DESCRIPTION
**Pull request by cavcrosby**
@stgraber per our discussion. My implementations stick mostly to what you suggested but there are some subtle differences. I think documentation is at a good start.

Closes #9408

### CODE REVIEW

File: lxd/instance/drivers/driver_qemu.go [hunk 1]
```diff
@@ -1017,6 +1017,11 @@ func (d *qemu) Start(stateful bool) error {
 		return err
 	}
 
+	// Ensure secureboot is turned off for images that are not secureboot enabled
+	if d.localConfig["image.requirements.secureboot"] == "false" && shared.IsTrue(d.expandedConfig["security.secureboot"]) {
+		return errors.New("This image is not secure boot enabled")
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go:1022]: We prefer `fmt.Errorf("This image is not secure boot enabled")` for this

File: lxd/instance/drivers/driver_lxc.go [hunk 2]
```diff
@@ -1979,6 +1979,11 @@ func (d *lxc) startCommon() (string, []func() error, error) {
 		return "", nil, errors.Wrap(err, "Load go-lxc struct")
 	}
 
+	// Ensure cgroup v1 configuration is set appropriately with the image using systemd
+	if d.localConfig["image.requirements.cgroup"] == "v1" && !shared.PathExists("/sys/fs/cgroup/systemd") {
+		return "", nil, errors.New("Failed to start container: requires host to use cgroups v1")
```
Comment (stgraber) [inline lxd/instance/drivers/driver_lxc.go:1984]: We prefer `fmt.Errorf("Failed to start container: requires host to use cgroups v1")` for this

File: lxd/instance/drivers/driver_qemu.go [hunk 3]
```diff
@@ -1017,6 +1017,11 @@ func (d *qemu) Start(stateful bool) error {
 		return err
 	}
 
+	// Ensure secureboot is turned off for images that are not secureboot enabled
+	if d.localConfig["image.requirements.secureboot"] == "false" && shared.IsTrue(d.expandedConfig["security.secureboot"]) {
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go:1021]: An empty `security.secureboot` means that it's enabled, so you need to check for empty too.

Comment (stgraber) [reply lxd/instance/drivers/driver_qemu.go:1021]: So something like `(d.expandedConfig["security.secureboot"] == "" || shared.IsTrue(d.expandedConfig["security.secureboot"]))` should do it

**Comment by stgraber** [comment]
Looks good, just a few tiny tweaks.

I think we should also add an API extension for this. For that, add an extra commit titled `api: image_restrictions` and have that contain an addition to both:
 - doc/api-extensions.md
 - shared/version/api.go

File: lxd/instance/drivers/driver_qemu.go [hunk 4]
```diff
@@ -1018,8 +1018,8 @@ func (d *qemu) Start(stateful bool) error {
 	}
 
 	// Ensure secureboot is turned off for images that are not secureboot enabled
-	if d.localConfig["image.requirements.secureboot"] == "false" && shared.IsTrue(d.expandedConfig["security.secureboot"]) {
-		return errors.New("This image is not secure boot enabled")
+	if d.localConfig["image.requirements.secureboot"] == "false" && (d.expandedConfig["security.secureboot"] == "" || shared.IsTrue(d.expandedConfig["security.secureboot"])) {
```
Comment (cavcrosby) [inline lxd/instance/drivers/driver_qemu.go:1021]: Err, secureboot logical expression was fixed too. The commit title is only telling half the story.

Comment (cavcrosby) [reply lxd/instance/drivers/driver_qemu.go:1021]: Otherwise this should be good to go.

**Comment by stgraber** [comment]
Thanks!

I'll do a quick review again, then will rebase to fix that conflict and merge your fixes into their respective commits.
Then just wait for Jenkins to give the all good and this can get merged!

File: lxd/instance/drivers/driver_qemu.go [hunk 5]
```diff
@@ -1017,6 +1017,11 @@ func (d *qemu) Start(stateful bool) error {
 		return err
 	}
 
+	// Ensure secureboot is turned off for images that are not secureboot enabled
+	if d.localConfig["image.requirements.secureboot"] == "false" && (d.expandedConfig["security.secureboot"] == "" || shared.IsTrue(d.expandedConfig["security.secureboot"])) {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1021]: Please can you use !shared.IsTrue

File: shared/simplestreams/products.go [hunk 6]
```diff
@@ -25,6 +25,7 @@ type Product struct {
 	Aliases         string                    `json:"aliases"`
 	Architecture    string                    `json:"arch"`
 	OperatingSystem string                    `json:"os"`
+	LxdRequirements map[string]string         `json:"lxd_requirements,omitempty"`
```
Comment (tomponline) [inline shared/simplestreams/products.go:28]: Should be LXDRequirements

File: lxd/instance/drivers/driver_qemu.go (hunk 5 above)
Comment (stgraber) [reply lxd/instance/drivers/driver_qemu.go:1021]: Hmm, no, that won't work.
If you don't have `security.secureboot` set, the default is `true` but if you do `!shared.IsTrue` that empty string will be causing `IsTrue` to return `false` skipping the test.

File: shared/simplestreams/products.go (hunk 6 above)
Comment (stgraber) [reply shared/simplestreams/products.go:28]: Will fix that one during rebase.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/instance/drivers/driver_qemu.go (hunk 5 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1021]: What I mean is we shouldn't be comparing to "false" literal unless we also compare to "true" literal for consistency.

Comment (stgraber) [reply lxd/instance/drivers/driver_qemu.go:1021]: Ah, I was looking at the other half of the check :)

Comment (stgraber) [reply lxd/instance/drivers/driver_qemu.go]: Added `shared.IsFalse` which we can then use for this test

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Thanks!

File: lxd/instance/drivers/driver_qemu.go (hunk 5 above)
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go]: Thanks

**Pull request merged**